flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +128,293 @@ public DefaultScheduler(
+		final Set<ExecutionVertexID> verticesToDeploy = deploymentOptionsByVertex.keySet();
+		final Map<ExecutionVertexID, ExecutionVertexVersion> requiredVersionByVertex = executionVertexVersioner.recordVertexModifications(verticesToDeploy);
+
+		prepareToDeployVertices(verticesToDeploy);
+
+		final Collection<SlotExecutionVertexAssignment> slotExecutionVertexAssignments = allocateSlots(executionVertexDeploymentOptions);
+
+		final Collection<DeploymentHandle> deploymentHandles = createDeploymentHandles(
+			requiredVersionByVertex,
+			deploymentOptionsByVertex,
+			slotExecutionVertexAssignments);
+
+		if (isDeployIndividually()) {
+			deployIndividually(deploymentHandles);
+		} else {
+			waitForAllSlotsAndDeploy(deploymentHandles);
+		}
+	}
+
+	private static Map<ExecutionVertexID, ExecutionVertexDeploymentOption> groupDeploymentOptionsByVertexId(
+			final Collection<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions) {
+		return executionVertexDeploymentOptions.stream().collect(Collectors.toMap(
+				ExecutionVertexDeploymentOption::getExecutionVertexId,
+				Function.identity()));
+	}
+
+	private void prepareToDeployVertices(final Set<ExecutionVertexID> verticesToDeploy) {
+		cancelSlotAssignments(verticesToDeploy);
+		resetForNewExecution(verticesToDeploy);
+		transitionToScheduled(verticesToDeploy);
+	}
+
+	private void cancelSlotAssignments(final Collection<ExecutionVertexID> vertices) {
+		vertices.forEach(executionVertexId -> executionSlotAllocator.cancel(executionVertexId));
+	}
+
+	private Collection<SlotExecutionVertexAssignment> allocateSlots(final Collection<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions) {
+		return executionSlotAllocator.allocateSlotsFor(executionVertexDeploymentOptions
+			.stream()
+			.map(ExecutionVertexDeploymentOption::getExecutionVertexId)
+			.map(this::getExecutionVertex)
+			.map(ExecutionVertexSchedulingRequirementsMapper::from)
+			.collect(Collectors.toList()));
+	}
+
+	private static Collection<DeploymentHandle> createDeploymentHandles(
+		final Map<ExecutionVertexID, ExecutionVertexVersion> requiredVersionByVertex,
+		final Map<ExecutionVertexID, ExecutionVertexDeploymentOption> deploymentOptionsByVertex,
+		final Collection<SlotExecutionVertexAssignment> slotExecutionVertexAssignments) {
+
+		return slotExecutionVertexAssignments
+			.stream()
+			.map(slotExecutionVertexAssignment -> {
+				final ExecutionVertexID executionVertexId = slotExecutionVertexAssignment.getExecutionVertexId();
+				return new DeploymentHandle(
+					requiredVersionByVertex.get(executionVertexId),
+					deploymentOptionsByVertex.get(executionVertexId),
+					slotExecutionVertexAssignment);
+			})
+			.collect(Collectors.toList());
+	}
+
+	private boolean isDeployIndividually() {
+		return schedulingStrategy instanceof LazyFromSourcesSchedulingStrategy;
```

Comment:
I think this check is not friendly for future scheduling strategies.

Maybe we should define SchedulerOperations#allocateSlotsAndDeploy as batch scheduling and all vertices in a batch should allocate resources and get deployed in a batch.

LazyFromSourcesSchedulingStrategy can avoid batch scheduling by invoking SchedulerOperations#allocateSlotsAndDeploy for each task to schedule.

Reply:
We can do it in a follow up task to not block `DefaultScheduler` from getting to a working state.

Reply:
> I think this check is not friendly for future scheduling strategies.

I agree. `deployIndividually` has the advantage that one can run batch jobs with higher parallelism than there are slots (this would not be possible in the `waitForAllSlotsAndDeploy` code path). In theory it is possible to have a single code path for both eager and lazy scheduling. However, I did not have time to think it through. I think this requires a follow up task.

Reply:
I opened a ticket FLINK-14162 to follow up. Let's skip it for now and do it after the DefaultScheduler is merged.